File: packages/python/pii_vault/hosted_vault.py
from typing import Optional

import httpx


class HostedVault:
    """
    Cloud-backed vault — stores token↔value mappings in the Mawlaia hosted API.

    Usage::

        from pii_vault import SafeOpenAI, HostedVault

        client = SafeOpenAI(
            api_key="sk-...",
            vault_key="my-hmac-secret",
            vault=HostedVault(api_key="mwl_live_..."),
        )

    Values are encrypted at rest using per-user Fernet keys on the server.
    In transit they travel over HTTPS only.
    """

    DEFAULT_URL = "https://api.mawlaia.com"

    def __init__(self, api_key: str, vault_url: str = DEFAULT_URL, timeout: float = 10.0):
        self._base = vault_url.rstrip("/")
        self._headers = {"Authorization": f"Bearer {api_key}", "Content-Type": "application/json"}
        self._client = httpx.Client(headers=self._headers, timeout=timeout)
        self._pending: list[dict] = []  # batched before flush()

    # ── Batch store (called by Tokenizer) ─────────────────────────────────────

    def store(
        self,
        token:       str,
        value:       str,
        entity_type: str,
        subject_id:  Optional[str] = None,
    ) -> None:
        """Queue an entry for batch upload. Call flush() after tokenize() completes."""
        self._pending.append({
            "token":       token,
            "value":       value,
            "entity_type": entity_type,
            "subject_id":  subject_id,
        })

    def flush(self) -> int:
        """Upload all pending entries to the hosted vault in one API call."""
        if not self._pending:
            return 0
        resp = self._client.post(
            f"{self._base}/v1/pii-vault/tokens",
            json={"entries": self._pending},
        )
        resp.raise_for_status()
        self._pending.clear()
        return resp.json().get("stored", 0)

    # ── Retrieve (called by Tokenizer) ────────────────────────────────────────

    def retrieve(self, token: str) -> Optional[str]:
        values = self.batch_retrieve([token])
        return values.get(token)

    def batch_retrieve(self, tokens: list[str]) -> dict[str, Optional[str]]:
        if not tokens:
            return {}
        resp = self._client.post(
            f"{self._base}/v1/pii-vault/tokens/lookup",
            json={"tokens": tokens},
        )
        resp.raise_for_status()
        return resp.json().get("values", {})

    # ── DSAR ──────────────────────────────────────────────────────────────────

    def delete_subject(self, subject_id: str) -> int:
        resp = self._client.delete(f"{self._base}/v1/pii-vault/subjects/{subject_id}")
        resp.raise_for_status()
        return resp.json().get("deleted_count", 0)

    def list_subject(self, subject_id: str) -> list[dict]:
        resp = self._client.get(f"{self._base}/v1/pii-vault/subjects/{subject_id}")
        resp.raise_for_status()
        return resp.json().get("tokens", [])

    def count(self) -> int:
        resp = self._client.get(f"{self._base}/v1/pii-vault/stats")
        resp.raise_for_status()
        return resp.json().get("total_entries", 0)

    # ── Hosted tokenize / detokenize (FPE) ────────────────────────────────────

    def tokenize_text(
        self,
        text: str,
        entities: list[str] | None = None,
        format_preserving: bool = True,
        subject_id: str | None = None,
    ) -> dict:
        """
        Detect PII in *text* server-side and replace with format-preserving fakes.
        Returns {"text": <de-identified>, "entities": [...], "stored": int}.
        """
        payload: dict = {"text": text, "format_preserving": format_preserving}
        if entities:
            payload["entities"] = entities
        if subject_id:
            payload["subject_id"] = subject_id
        resp = self._client.post(f"{self._base}/v1/pii-vault/tokenize", json=payload)
        resp.raise_for_status()
        return resp.json()

    def detokenize_text(self, text: str) -> str:
        """Reverse a previously tokenized text, restoring original PII values."""
        resp = self._client.post(f"{self._base}/v1/pii-vault/detokenize", json={"text": text})
        resp.raise_for_status()
        return resp.json().get("text", text)

    def __del__(self):
        try:
            self._client.close()
        except Exception:
            pass

File: packages/typescript/src/hosted_vault.ts
export interface HostedVaultOptions {
  apiKey: string;
  vaultUrl?: string;
  timeout?: number;
}

interface PendingEntry {
  token: string;
  value: string;
  entity_type: string;
  subject_id?: string;
}

export class HostedVault {
  private readonly base: string;
  private readonly headers: Record<string, string>;
  private readonly timeout: number;
  private pending: PendingEntry[] = [];

  constructor({ apiKey, vaultUrl = "https://api.mawlaia.com", timeout = 10_000 }: HostedVaultOptions) {
    this.base = vaultUrl.replace(/\/$/, "");
    this.headers = {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/json",
    };
    this.timeout = timeout;
  }

  private async post<T>(path: string, body: unknown): Promise<T> {
    const controller = new AbortController();
    const id = setTimeout(() => controller.abort(), this.timeout);
    try {
      const res = await fetch(`${this.base}${path}`, {
        method: "POST",
        headers: this.headers,
        body: JSON.stringify(body),
        signal: controller.signal,
      });
      if (!res.ok) throw new Error(`Mawlaia API error: ${res.status} ${await res.text()}`);
      return res.json() as Promise<T>;
    } finally {
      clearTimeout(id);
    }
  }

  private async get<T>(path: string): Promise<T> {
    const controller = new AbortController();
    const id = setTimeout(() => controller.abort(), this.timeout);
    try {
      const res = await fetch(`${this.base}${path}`, {
        headers: this.headers,
        signal: controller.signal,
      });
      if (!res.ok) throw new Error(`Mawlaia API error: ${res.status} ${await res.text()}`);
      return res.json() as Promise<T>;
    } finally {
      clearTimeout(id);
    }
  }

  store(token: string, value: string, entityType: string, subjectId?: string): void {
    this.pending.push({ token, value, entity_type: entityType, subject_id: subjectId });
  }

  async flush(): Promise<number> {
    if (this.pending.length === 0) return 0;
    const entries = this.pending.splice(0);
    const data = await this.post<{ stored: number }>("/v1/pii-vault/tokens", { entries });
    return data.stored;
  }

  async retrieve(token: string): Promise<string | null> {
    const map = await this.batchRetrieve([token]);
    return map[token] ?? null;
  }

  async batchRetrieve(tokens: string[]): Promise<Record<string, string | null>> {
    if (tokens.length === 0) return {};
    const data = await this.post<{ values: Record<string, string | null> }>("/v1/pii-vault/tokens/lookup", { tokens });
    return data.values;
  }

  async deleteSubject(subjectId: string): Promise<number> {
    const controller = new AbortController();
    const id = setTimeout(() => controller.abort(), this.timeout);
    try {
      const res = await fetch(`${this.base}/v1/pii-vault/subjects/${subjectId}`, {
        method: "DELETE",
        headers: this.headers,
        signal: controller.signal,
      });
      if (!res.ok) throw new Error(`Mawlaia API error: ${res.status}`);
      const data = await res.json() as { deleted_count: number };
      return data.deleted_count;
    } finally {
      clearTimeout(id);
    }
  }

  async listSubject(subjectId: string): Promise<Record<string, unknown>[]> {
    const data = await this.get<{ tokens: Record<string, unknown>[] }>(`/v1/pii-vault/subjects/${subjectId}`);
    return data.tokens;
  }

  async count(): Promise<number> {
    const data = await this.get<{ total_entries: number }>("/v1/pii-vault/stats");
    return data.total_entries;
  }

  async tokenizeText(
    text: string,
    options: { entities?: string[]; formatPreserving?: boolean; subjectId?: string } = {},
  ): Promise<{ text: string; entities: Record<string, unknown>[]; stored: number }> {
    const body: Record<string, unknown> = {
      text,
      format_preserving: options.formatPreserving ?? true,
    };
    if (options.entities) body.entities = options.entities;
    if (options.subjectId) body.subject_id = options.subjectId;
    return this.post("/v1/pii-vault/tokenize", body);
  }

  async detokenizeText(text: string): Promise<string> {
    const data = await this.post<{ text: string }>("/v1/pii-vault/detokenize", { text });
    return data.text;
  }
}
